pkg/function/client/options.go

```diff
@@ -0,0 +1,27 @@
+package client
+
+import (
+	"github.com/numaproj/numaflow-go/pkg/apis/proto/function/v1/funcmock"
+)
+
+type options struct {
+	mockClient gRPClientOption
+}
+
+type Option interface {
```

Comment:
Use the approach like https://github.com/numaproj/numaflow/blob/main/pkg/shared/clients/jetstream/options.go#L23, which is simpler.

Reply:
@vigith suggested https://github.com/uber-go/guide/blob/master/style.md#functional-options
I'm ok with either one

Reply:
The reason I asked for struct type here is so that we can attach other complex functions to the data type later. These are not the usual options, but one can imagine these like `GetXXX` and `SetXXX`.

Reply:
I am thinking of using `opts ...MapOptions` available in the `handle` function. Eg., one could do `opts.GetWatermark()`.

Reply:
looks like these are `options` to gprc. This can be implemented any way we like. My comments are mostly for `pkg/function/options.go`

Reply:
How to make it happen in the `handle` function? I could imagine the function looks like `handle(key string, value []byte, meta metadata)`, is there a way to do it like `handle(key string, value []byte, opts ...Options)`, where `opts` is an array of interface?

Reply:
@whynowy I have this experimental code for handler function options
https://github.com/numaproj/numaflow-go/tree/grpc-option

The idea is to first set up the MapOption when registering the mapper.
Then, in the DoFn, the MapOption is converted to a HandleOption which is passed to the HandleDo().

Reply:
> @whynowy I have this experimental code for handler function options https://github.com/numaproj/numaflow-go/tree/grpc-option
> 
> The idea is to first set up the MapOption when registering the mapper. Then, in the DoFn, the MapOption is converted to a HandleOption which is passed to the HandleDo().

I understand that. 

The developer/user should not be responsible for populate those options in their code, instead, what should be done is, the SDK populate them somewhere (e.g. a struct object), and the developer is only supposed to call get functions.

Reply:
The following part of code should be not written by each of the SDK users.

```
options := functionsdk.MapOptions{
		EventTime: time.Time{},
	}
	for _, o := range inputOptions {
		o.Get(&options)
	}
```

Reply:
I've updated the func sdk client option implementation

Reply:
Another approach specifically for Golang is, get those stuff into `context`, for example, we have a struct to wrap all of those options, instead of putting the struct in the function signature, put it in the context.

For Python, I guess the common practice is having a `metadata` in the function signature.